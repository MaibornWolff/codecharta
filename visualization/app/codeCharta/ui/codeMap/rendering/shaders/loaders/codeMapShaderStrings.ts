export class CodeMapShaderStrings {
	vertexShaderCode: string
	fragmentShaderCode: string

	constructor() {
		this.vertexShaderCode = require("../glsl/codeMapVertexShader.glsl")
		this.fragmentShaderCode = require("../glsl/codeMapFragmentShader.glsl")
	}
}
