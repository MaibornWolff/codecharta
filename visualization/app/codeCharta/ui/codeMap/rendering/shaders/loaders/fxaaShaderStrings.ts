import { Texture, Vector2 } from "three"

interface uniforms {
	tDiffuse : {
		value : Texture,
	},
	resolution : {
		value : Vector2
	}
}

export class fxaaShaderStrings {
	vertexShader: string
	fragmentShader: string
	uniforms : uniforms

	constructor() {
		this.vertexShader = require("../glsl/fxaaVertexShader.glsl")
		this.fragmentShader = require("../glsl/fxaaFragmentShader.glsl")
		this.uniforms = {
			'tDiffuse': { value: null },
			'resolution': { value: new Vector2( 1 / 1024, 1 / 512 ) }
		}
	}
}
