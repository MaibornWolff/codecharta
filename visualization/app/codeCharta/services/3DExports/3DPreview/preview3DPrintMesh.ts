import { Mesh, ShaderMaterial, Vector3 } from "three"
import { Font } from "three/examples/jsm/loaders/FontLoader"
import { ColorRange, NodeMetricData } from "../../../codeCharta.model"
import { BaseplateMesh } from "./MeshModels/baseplateMesh"
import { GeneralMesh } from "./MeshModels/generalMesh"
import { MapMesh } from "./MeshModels/mapMesh"
import { BackPrintContainerMesh } from "./MeshModels/BackMeshModels/backPrintContainerMesh"
import { FrontPrintContainerMesh } from "./MeshModels/FrontMeshModels/frontPrintContainerMesh"
import font from "three/examples/fonts/helvetiker_regular.typeface.json"

export interface GeometryOptions {
    originalMapMesh: Mesh
    width: number
    areaMetricTitle: string
    areaMetricData: NodeMetricData
    heightMetricTitle: string
    heightMetricData: NodeMetricData
    colorMetricTitle: string
    colorMetricData: NodeMetricData
    colorRange: ColorRange
    frontText: string
    secondRowText: string
    qrCodeText: string
    defaultMaterial: ShaderMaterial
    numberOfColors: number
    layerHeight: number
    frontTextSize: number
    secondRowTextSize: number
    secondRowVisible: boolean
    printHeight: number
    mapSideOffset: number
    baseplateHeight: number
    logoSize: number
}

export class Preview3DPrintMesh {
    private printMesh: Mesh
    private currentSize: Vector3

    constructor(
        private geometryOptions: GeometryOptions,
        private frontPrintContainerMesh: FrontPrintContainerMesh = new FrontPrintContainerMesh(new Font(font)),
        private backPrintContainerMesh: BackPrintContainerMesh = new BackPrintContainerMesh(new Font(font)),
        private baseplateMesh: BaseplateMesh = new BaseplateMesh(),
        private mapMesh: MapMesh = new MapMesh()
    ) {
        this.printMesh = new Mesh()
        this.printMesh.name = "PrintMesh" //TODO: rename to PrintPreview
    }

    async initialize() {
        this.baseplateMesh = await this.baseplateMesh.init(this.geometryOptions)
        this.printMesh.add(this.baseplateMesh)

        this.mapMesh = await this.mapMesh.init(this.geometryOptions)
        this.printMesh.add(this.mapMesh)

        this.frontPrintContainerMesh = await this.frontPrintContainerMesh.init(this.geometryOptions)
        this.printMesh.add(this.frontPrintContainerMesh)

        this.backPrintContainerMesh = await this.backPrintContainerMesh.init(this.geometryOptions)
        this.printMesh.add(this.backPrintContainerMesh)

        this.calculateCurrentSize()
    }

    getThreeMesh(): Mesh {
        return this.printMesh
    }

    getMapMesh(): Mesh {
        return this.mapMesh
    }

    getSize(): Vector3 {
        return this.currentSize
    }

    async updateSize(wantedWidth: number): Promise<boolean> {
        this.geometryOptions.width = wantedWidth
        const oldWidth = this.currentSize.x

        await Promise.all(
            [...this.printMesh.children].map(async mesh => {
                if (mesh instanceof GeneralMesh && mesh.isGeneralSizeChangeMesh()) {
                    mesh.changeSize(this.geometryOptions, oldWidth)
                }
            })
        )

        this.calculateCurrentSize()
        return this.backPrintContainerMesh.isQRCodeVisible()
    }

    updateNumberOfColors(numberOfColors: number) {
        this.geometryOptions.numberOfColors = numberOfColors
        for (const mesh of this.printMesh.children) {
            if (mesh instanceof GeneralMesh) {
                mesh.updateColor(numberOfColors)
            }
        }
    }

    async addCustomLogo(dataUrl: string): Promise<void> {
        this.frontPrintContainerMesh.addCustomLogo(dataUrl, this.geometryOptions)
    }

    rotateCustomLogo() {
        this.frontPrintContainerMesh.rotateCustomLogo()
    }

    flipCustomLogo() {
        this.frontPrintContainerMesh.flipCustomLogo()
    }

    removeCustomLogo() {
        this.frontPrintContainerMesh.removeCustomLogo()
    }

    updateCustomLogoColor(newColor: string) {
        this.frontPrintContainerMesh.updateCustomLogoColor(newColor)
    }

    updateFrontText(frontText: string) {
        this.frontPrintContainerMesh.updateFrontText(frontText, this.geometryOptions)
    }

    updateSecondRowVisibility(isSecondRowVisible: boolean) {
        if (this.geometryOptions.secondRowVisible === isSecondRowVisible) {
            return
        }
        this.geometryOptions.secondRowVisible = isSecondRowVisible

        this.frontPrintContainerMesh.updateSecondRowVisibility(this.geometryOptions)
        this.baseplateMesh.changeSize(this.geometryOptions)
    }

    async updateSecondRowText(secondRowText: string) {
        await this.frontPrintContainerMesh.updateSecondRowText(secondRowText, this.geometryOptions)
    }

    private calculateCurrentSize() {
        const currentWidth = this.baseplateMesh.getWidth()
        const currentDepth = this.baseplateMesh.getDepth()
        const currentHeight = this.baseplateMesh.getHeight() + this.mapMesh.getHeight()
        this.currentSize = new Vector3(currentWidth, currentDepth, currentHeight)
    }

    async updateQrCodeText(qrCodeText: string): Promise<void> {
        await this.backPrintContainerMesh.updateQrCodeText(qrCodeText, this.geometryOptions)
    }

    updateQrCodeVisibility(qrCodeVisible: boolean) {
        this.backPrintContainerMesh.updateQrCodeVisibility(qrCodeVisible)
    }
}

export function calculateMaxPossibleWidthForPreview3DPrintMesh(
    maxSize: Vector3,
    mapMesh: Mesh,
    frontTextSize: number,
    baseplateHeight: number,
    mapSideOffset: number
): number {
    const printerWidth = maxSize.x
    const printerDepth = maxSize.y
    const printerHeight = maxSize.z

    const widthFromWidth = printerWidth
    const widthFromDepth = printerDepth - frontTextSize
    if (!mapMesh.geometry.boundingBox) {
        mapMesh.geometry.computeBoundingBox()
    }
    const mapCurrentHeight = mapMesh.geometry.boundingBox.max.z - mapMesh.geometry.boundingBox.min.z
    const widthFromHeight = ((printerHeight - baseplateHeight) * mapMesh.geometry.boundingBox.max.x) / mapCurrentHeight + 2 * mapSideOffset

    return Math.min(widthFromWidth, widthFromDepth, widthFromHeight)
}
