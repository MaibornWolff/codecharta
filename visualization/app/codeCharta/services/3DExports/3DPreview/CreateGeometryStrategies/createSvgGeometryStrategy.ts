import { BufferGeometry, ExtrudeGeometry } from "three"
import { SVGLoader } from "three/examples/jsm/loaders/SVGLoader"
import { CreateGeometryStrategy, CreateGeometryStrategyOptions } from "./createGeometryStrategy"
import { GeometryOptions } from "../preview3DPrintMesh"
import * as BufferGeometryUtils from "three/examples/jsm/utils/BufferGeometryUtils.js"

export interface CreateSvgGeometryStrategyOptions extends CreateGeometryStrategyOptions {
    filePath: string
    size: number
    side: "front" | "back"
}
export class CreateSvgGeometryStrategy implements CreateGeometryStrategy {
    async create(
        geometryOptions: GeometryOptions,
        createSvgGeometryStrategyOptions: CreateSvgGeometryStrategyOptions
    ): Promise<BufferGeometry> {
        const filePath = createSvgGeometryStrategyOptions.filePath
        const loader = new SVGLoader()
        return new Promise((resolve, reject) => {
            loader.load(
                filePath,
                function (data) {
                    const paths = data.paths
                    const geometries: BufferGeometry[] = []

                    for (const path of paths) {
                        const shapes = path.toShapes(true)

                        for (const shape of shapes) {
                            const geometry = new ExtrudeGeometry(shape, {
                                depth: geometryOptions.printHeight,
                                bevelEnabled: false
                            })
                            geometries.push(geometry)
                        }
                    }

                    const mergedGeometry = BufferGeometryUtils.mergeGeometries(geometries)

                    mergedGeometry.computeBoundingBox()
                    const width = mergedGeometry.boundingBox.max.x - mergedGeometry.boundingBox.min.x
                    const depth = mergedGeometry.boundingBox.max.y - mergedGeometry.boundingBox.min.y
                    const scale = createSvgGeometryStrategyOptions.size / Math.max(width, depth)
                    mergedGeometry.scale(scale, scale, 1)

                    mergedGeometry.center()
                    if (createSvgGeometryStrategyOptions.side === "back") {
                        mergedGeometry.rotateZ(Math.PI)
                    } else {
                        mergedGeometry.rotateZ(Math.PI)
                        mergedGeometry.rotateY(Math.PI)
                    }

                    resolve(mergedGeometry)
                },
                undefined,
                function (error) {
                    console.error(`Error loading ${filePath}`)
                    reject(error)
                }
            )
        })
    }
}
