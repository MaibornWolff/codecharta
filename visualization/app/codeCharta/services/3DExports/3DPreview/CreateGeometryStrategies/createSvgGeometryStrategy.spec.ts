import { SVGLoader } from "three/examples/jsm/loaders/SVGLoader"
import { CreateSvgGeometryStrategy, CreateSvgGeometryStrategyOptions } from "./createSvgGeometryStrategy"
import { GeometryOptions } from "../preview3DPrintMesh"

describe("CreateSvgGeometryStrategy", () => {
    let strategy: CreateSvgGeometryStrategy

    beforeEach(() => {
        strategy = new CreateSvgGeometryStrategy()
    })

    it("should create geometry from SVG and apply transformations", async () => {
        const geometryOptions: GeometryOptions = {
            printHeight: 10
        } as unknown as GeometryOptions

        const createSvgGeometryStrategyOptions: CreateSvgGeometryStrategyOptions = {
            filePath: "",
            size: 100,
            side: "front"
        } as unknown as CreateSvgGeometryStrategyOptions

        const svgData = `
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100">
                <path d="M10 10 H 90 V 90 H 10 L 10 10" />
            </svg>
        `

        const loader = new SVGLoader()
        const svg = loader.parse(svgData)
        jest.spyOn(SVGLoader.prototype, "load").mockImplementation((url, onLoad) => {
            onLoad(svg)
        })

        const geometry = await strategy.create(geometryOptions, createSvgGeometryStrategyOptions)

        expect(geometry.attributes).toMatchSnapshot()
    })

    it("should create geometry from SVG and apply transformations for back side", async () => {
        const geometryOptions: GeometryOptions = {
            printHeight: 10
        } as unknown as GeometryOptions

        const createSvgGeometryStrategyOptions: CreateSvgGeometryStrategyOptions = {
            filePath: "",
            size: 100,
            side: "back"
        } as unknown as CreateSvgGeometryStrategyOptions

        const svgData = `
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100">
                <path d="M10 10 H 90 V 90 H 10 L 10 10" />
            </svg>
        `

        const loader = new SVGLoader()
        const svg = loader.parse(svgData)
        jest.spyOn(SVGLoader.prototype, "load").mockImplementation((url, onLoad) => {
            onLoad(svg)
        })

        const geometry = await strategy.create(geometryOptions, createSvgGeometryStrategyOptions)

        expect(geometry.attributes).toMatchSnapshot()
    })
})
