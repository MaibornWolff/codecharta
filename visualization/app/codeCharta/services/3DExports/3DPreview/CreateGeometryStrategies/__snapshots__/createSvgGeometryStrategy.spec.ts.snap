// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateSvgGeometryStrategy should create geometry from SVG and apply transformations 1`] = `
{
  "normal": {
    "array": [
      -1.2246468525851679e-16,
      0,
      1,
      -1.2246468525851679e-16,
      0,
      1,
      -1.2246468525851679e-16,
      0,
      1,
      -1.2246468525851679e-16,
      0,
      1,
      -1.2246468525851679e-16,
      0,
      1,
      -1.2246468525851679e-16,
      0,
      1,
      1.2246468525851679e-16,
      0,
      -1,
      1.2246468525851679e-16,
      0,
      -1,
      1.2246468525851679e-16,
      0,
      -1,
      1.2246468525851679e-16,
      0,
      -1,
      1.2246468525851679e-16,
      0,
      -1,
      1.2246468525851679e-16,
      0,
      -1,
      1,
      1.2246468525851679e-16,
      1.2246468525851679e-16,
      1,
      1.2246468525851679e-16,
      1.2246468525851679e-16,
      1,
      1.2246468525851679e-16,
      1.2246468525851679e-16,
      1,
      1.2246468525851679e-16,
      1.2246468525851679e-16,
      1,
      1.2246468525851679e-16,
      1.2246468525851679e-16,
      1,
      1.2246468525851679e-16,
      1.2246468525851679e-16,
      1.2246468525851679e-16,
      -1,
      1.499759904157988e-32,
      1.2246468525851679e-16,
      -1,
      1.499759904157988e-32,
      1.2246468525851679e-16,
      -1,
      1.499759904157988e-32,
      1.2246468525851679e-16,
      -1,
      1.499759904157988e-32,
      1.2246468525851679e-16,
      -1,
      1.499759904157988e-32,
      1.2246468525851679e-16,
      -1,
      1.499759904157988e-32,
      -1,
      -1.2246468525851679e-16,
      -1.2246468525851679e-16,
      -1,
      -1.2246468525851679e-16,
      -1.2246468525851679e-16,
      -1,
      -1.2246468525851679e-16,
      -1.2246468525851679e-16,
      -1,
      -1.2246468525851679e-16,
      -1.2246468525851679e-16,
      -1,
      -1.2246468525851679e-16,
      -1.2246468525851679e-16,
      -1,
      -1.2246468525851679e-16,
      -1.2246468525851679e-16,
      -1.2246468525851679e-16,
      1,
      -1.499759904157988e-32,
      -1.2246468525851679e-16,
      1,
      -1.499759904157988e-32,
      -1.2246468525851679e-16,
      1,
      -1.499759904157988e-32,
      -1.2246468525851679e-16,
      1,
      -1.499759904157988e-32,
      -1.2246468525851679e-16,
      1,
      -1.499759904157988e-32,
      -1.2246468525851679e-16,
      1,
      -1.499759904157988e-32,
    ],
    "itemSize": 3,
    "normalized": false,
    "type": "Float32Array",
  },
  "position": {
    "array": [
      50,
      50,
      5,
      -50,
      50,
      5,
      -50,
      -50,
      5,
      -50,
      -50,
      5,
      50,
      -50,
      5,
      50,
      50,
      5,
      -50,
      -50,
      -5,
      -50,
      50,
      -5,
      50,
      50,
      -5,
      50,
      50,
      -5,
      50,
      -50,
      -5,
      -50,
      -50,
      -5,
      50,
      50,
      5,
      50,
      -50,
      5,
      50,
      50,
      -5,
      50,
      -50,
      5,
      50,
      -50,
      -5,
      50,
      50,
      -5,
      50,
      -50,
      5,
      -50,
      -50,
      5,
      50,
      -50,
      -5,
      -50,
      -50,
      5,
      -50,
      -50,
      -5,
      50,
      -50,
      -5,
      -50,
      -50,
      5,
      -50,
      50,
      5,
      -50,
      -50,
      -5,
      -50,
      50,
      5,
      -50,
      50,
      -5,
      -50,
      -50,
      -5,
      -50,
      50,
      5,
      50,
      50,
      5,
      -50,
      50,
      -5,
      50,
      50,
      5,
      50,
      50,
      -5,
      -50,
      50,
      -5,
    ],
    "itemSize": 3,
    "normalized": false,
    "type": "Float32Array",
  },
  "uv": {
    "array": [
      90,
      10,
      10,
      10,
      10,
      90,
      10,
      90,
      90,
      90,
      90,
      10,
      10,
      90,
      10,
      10,
      90,
      10,
      90,
      10,
      90,
      90,
      10,
      90,
      10,
      1,
      90,
      1,
      10,
      -9,
      90,
      1,
      90,
      -9,
      10,
      -9,
      90,
      1,
      10,
      1,
      90,
      -9,
      10,
      1,
      10,
      -9,
      90,
      -9,
      90,
      1,
      10,
      1,
      90,
      -9,
      10,
      1,
      10,
      -9,
      90,
      -9,
      10,
      1,
      90,
      1,
      10,
      -9,
      90,
      1,
      90,
      -9,
      10,
      -9,
    ],
    "itemSize": 2,
    "normalized": false,
    "type": "Float32Array",
  },
}
`;

exports[`CreateSvgGeometryStrategy should create geometry from SVG and apply transformations for back side 1`] = `
{
  "normal": {
    "array": [
      -0,
      0,
      -1,
      -0,
      0,
      -1,
      -0,
      0,
      -1,
      0,
      0,
      -1,
      0,
      0,
      -1,
      0,
      0,
      -1,
      0,
      0,
      1,
      0,
      0,
      1,
      0,
      0,
      1,
      0,
      0,
      1,
      0,
      0,
      1,
      0,
      0,
      1,
      -1,
      1.2246468525851679e-16,
      0,
      -1,
      1.2246468525851679e-16,
      0,
      -1,
      1.2246468525851679e-16,
      0,
      -1,
      1.2246468525851679e-16,
      0,
      -1,
      1.2246468525851679e-16,
      0,
      -1,
      1.2246468525851679e-16,
      0,
      -1.2246468525851679e-16,
      -1,
      0,
      -1.2246468525851679e-16,
      -1,
      0,
      -1.2246468525851679e-16,
      -1,
      0,
      -1.2246468525851679e-16,
      -1,
      0,
      -1.2246468525851679e-16,
      -1,
      0,
      -1.2246468525851679e-16,
      -1,
      0,
      1,
      -1.2246468525851679e-16,
      0,
      1,
      -1.2246468525851679e-16,
      0,
      1,
      -1.2246468525851679e-16,
      0,
      1,
      -1.2246468525851679e-16,
      0,
      1,
      -1.2246468525851679e-16,
      0,
      1,
      -1.2246468525851679e-16,
      0,
      1.2246468525851679e-16,
      1,
      0,
      1.2246468525851679e-16,
      1,
      0,
      1.2246468525851679e-16,
      1,
      0,
      1.2246468525851679e-16,
      1,
      -0,
      1.2246468525851679e-16,
      1,
      -0,
      1.2246468525851679e-16,
      1,
      -0,
    ],
    "itemSize": 3,
    "normalized": false,
    "type": "Float32Array",
  },
  "position": {
    "array": [
      -50,
      50,
      -5,
      50,
      50,
      -5,
      50,
      -50,
      -5,
      50,
      -50,
      -5,
      -50,
      -50,
      -5,
      -50,
      50,
      -5,
      50,
      -50,
      5,
      50,
      50,
      5,
      -50,
      50,
      5,
      -50,
      50,
      5,
      -50,
      -50,
      5,
      50,
      -50,
      5,
      -50,
      50,
      -5,
      -50,
      -50,
      -5,
      -50,
      50,
      5,
      -50,
      -50,
      -5,
      -50,
      -50,
      5,
      -50,
      50,
      5,
      -50,
      -50,
      -5,
      50,
      -50,
      -5,
      -50,
      -50,
      5,
      50,
      -50,
      -5,
      50,
      -50,
      5,
      -50,
      -50,
      5,
      50,
      -50,
      -5,
      50,
      50,
      -5,
      50,
      -50,
      5,
      50,
      50,
      -5,
      50,
      50,
      5,
      50,
      -50,
      5,
      50,
      50,
      -5,
      -50,
      50,
      -5,
      50,
      50,
      5,
      -50,
      50,
      -5,
      -50,
      50,
      5,
      50,
      50,
      5,
    ],
    "itemSize": 3,
    "normalized": false,
    "type": "Float32Array",
  },
  "uv": {
    "array": [
      90,
      10,
      10,
      10,
      10,
      90,
      10,
      90,
      90,
      90,
      90,
      10,
      10,
      90,
      10,
      10,
      90,
      10,
      90,
      10,
      90,
      90,
      10,
      90,
      10,
      1,
      90,
      1,
      10,
      -9,
      90,
      1,
      90,
      -9,
      10,
      -9,
      90,
      1,
      10,
      1,
      90,
      -9,
      10,
      1,
      10,
      -9,
      90,
      -9,
      90,
      1,
      10,
      1,
      90,
      -9,
      10,
      1,
      10,
      -9,
      90,
      -9,
      10,
      1,
      90,
      1,
      10,
      -9,
      90,
      1,
      90,
      -9,
      10,
      -9,
    ],
    "itemSize": 2,
    "normalized": false,
    "type": "Float32Array",
  },
}
`;
